pkg/sheets/sheets.go

```diff
@@ -1,66 +1,154 @@
 	}
+
+	// deserialize the data file
+	var sheets []Sheet
+	if err := json.NewDecoder(&dest.Buffer).Decode(&sheets); err != nil {
+		logger.WithError(err).Error("json.Decode() failed")
+		return nil
+	}
+	return sheets
 }
 
-func NewSheetFromRequest(r *http.Request) (Sheet, error) {
-	partNumber, _ := strconv.Atoi(r.FormValue("part_number"))
-	sheet := Sheet{
-		Project:    r.FormValue("project"),
-		Instrument: r.FormValue("instrument"),
-		PartNumber: partNumber,
+func (x Sheets) Store(ctx context.Context, sheets []Sheet, fileBytes []byte) bool {
+	// first, validate all the sheets
+	if ok := validateSheets(sheets); !ok {
+		return false
+	}
+
+	// hash the pdf bytes
+	fileKey := fmt.Sprintf("%x.pdf", md5.Sum(fileBytes))
+	for i := range sheets {
+		sheets[i].FileKey = fileKey
+	}
+
+	// store the pdf
+	x.PutObject(fileKey, &storage.Object{
+		ContentType: "application/pdf",
+		Buffer:      *bytes.NewBuffer(fileBytes),
+	})
+
+	// now we update the data file
+	// read+modify+write means we need a lock
+	if ok := x.Lock(ctx); !ok {
+		return false
 	}
-	return sheet, sheet.Validate()
+	defer x.Unlock()
+
+	// pull down the sheets data
+	allSheets := x.List()
+	if allSheets == nil {
+		return false
+	}
+	allSheets = append(allSheets, sheets...)
+
+	// encode the data file
+	var buffer bytes.Buffer
+	if err := json.NewEncoder(&buffer).Encode(&allSheets); err != nil {
+		logger.WithError(err).Error("json.Encode() failed")
+		return false
+	}
+
+	// store the data file with a revision
+	backupName := fmt.Sprintf("%s-%s", DataFile, time.Now().UTC().Format(time.RFC3339))
+	if ok := x.PutObject(backupName, &storage.Object{
+		ContentType: "application/json",
+		Buffer:      buffer,
+	}); !ok {
+		return false
+	}
+
+	// write the data file
+	return x.PutObject(DataFile, &storage.Object{
+		ContentType: "application/json",
+		Buffer:      buffer,
+	})
 }
 
-func (x Sheet) ObjectKey() string {
-	return fmt.Sprintf("%s-%s-%d.pdf", x.Project, x.Instrument, x.PartNumber)
+func validateSheets(sheets []Sheet) bool {
+	for _, sheet := range sheets {
+		if err := sheet.Validate(); err != nil {
+			logger.WithError(err).Error("sheet failed validation")
+			return false
+		}
+	}
+	return true
 }
 
-func (x Sheet) Link() string {
-	return fmt.Sprintf("/download?bucket=%s&key=%s", BucketName, x.ObjectKey())
+type Sheet struct {
+	Project    string `json:"project"`
+	PartName   string `json:"part_name"`
+	PartNumber uint8  `json:"part_number"`
+	FileKey    string `json:"file_key"`
 }
 
-func (x Sheet) Tags() map[string]string {
-	return map[string]string{
-		"Project":     x.Project,
-		"Instrument":  x.Instrument,
-		"Part-Number": strconv.Itoa(x.PartNumber),
-	}
+func (x Sheet) String() string {
+	return fmt.Sprintf("Project: %s Part: %s-%d", x.Project, x.PartName, x.PartNumber)
+}
+
+func (x Sheet) ObjectKey() string {
+	return x.FileKey
+}
+
+func (x Sheet) Link(bucket string) string {
+	return fmt.Sprintf("/download?bucket=%s&object=%s", bucket, x.ObjectKey())
```

Comment:
Does this allow arbitrary reads?

Reply:
Access to this api is protected via basic auth. Additionally, the download handler will only provide download links for whitelisted buckets.